consensus/src/consensus_observer/observer.rs

```diff
@@ -264,8 +264,13 @@ impl ConsensusObserver {
             // Remove the committed blocks from the pending blocks
             remove_pending_blocks(pending_blocks, &ledger_info);
 
-            // Update the root ledger info
-            *root.lock() = ledger_info;
+            // Update the root ledger info. Note: we only want to do this if
+            // the new ledger info round is greater than the current root
+            // round. Otherwise, this can race with the state sync process.
+            let mut root = root.lock();
+            if ledger_info.commit_info().round() > root.commit_info().round() {
```

Comment:
I assume there's a guarantee somewhere that the epochs must be the same when the callback is called?

Reply:
Yes, good point 😄 It is implied. AFAICT, CO will only ever use state sync to cross the epoch boundary. 'cc @zekun000 to confirm.

But, let me harden this with an explicit check, too!

Reply:
I think today it probably relies on future epoch ledger to sync, but we should be able to make it cross epoch boundary by itself

Reply:
Cool, that matches my current understanding 😄

> but we should be able to make it cross epoch boundary by itself

SGTM -- sounds like post-MVP feature 🐒

Reply:
Added a check that the commit callback is for the same epoch 😄